### PR TITLE
Unit testing for api.js

### PR DESCRIPTION
**Pull request by kathyn262**
This pull request includes a

- [ ] Bug fix
- [x] New feature
- [x] Testing
- [ ] Translation

The following changes were made:
- Added api.test.js into integration subfolder

Comments
- This is the last file for issue #5. Can close issue #5 after merge

[#5](https://github.com/rithmschool/numbers_api/issues/5)

### CODE REVIEW

File: .github/ISSUE_TEMPLATE.md [hunk 1]
```diff
@@ -3,6 +3,7 @@ Thanks for your interest in contributing to this project. Please follow the Code
 
 Fill out this template with information that is relevant to the issue and if necessary provide information about how to replicate the issue so that we can understand the best way to approach the problem.
 -->
+
```
Comment (mmmaaatttttt) [inline .github/ISSUE_TEMPLATE.md:6]: this seems like an unrelated change. is it necessary?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -6,8 +6,8 @@ This pull request includes a
 
 The following changes were made
 
--
```
Comment (mmmaaatttttt) [inline .github/PULL_REQUEST_TEMPLATE.md:9]: changes to this file also seem unrelated. necessary?

File: CODE_OF_CONDUCT.md [hunk 3]
```diff
@@ -1,4 +1,3 @@
-
```
Comment (mmmaaatttttt) [inline CODE_OF_CONDUCT.md:1]: changes to this file also seem unrelated. necessary?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -7,5 +7,5 @@ Thanks for being willing to contribute!
 1. Ensure any install or build dependencies are removed before the end of the layer when doing a build.
```
Comment (mmmaaatttttt) [inline CONTRIBUTING.md:7]: changes to this file also seem unrelated. necessary?

File: __tests__/unit/integration/api.test.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
+const { appendToFile } = require("../../../routes/api");
```
Comment (mmmaaatttttt) [inline __tests__/unit/integration/api.test.js:1]: this folder location seems weird, shouldn't `integration` be a sibling directory of `unit` and not a child directory?

File: __tests__/unit/integration/api.test.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+      expect(data).toEqual("hello world");
+    });
+  });
+});
+
+describe("routes", () => {
+  // done() pattern needed for all tests
+  // without done() we get a jest TCPSERVERWRAP error
+
+  // test for route "/:num(-?[0-9]+)"
```
Comment (mmmaaatttttt) [inline __tests__/unit/integration/api.test.js:24]: you say this, but none of your tests use `done` :) using `async` / `await` should make `done` unnecessary. can you remove this comment?

File: __tests__/unit/integration/api.test.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+  test("1st route", async function () {
+    const res = await request(app).get("/1");
+    expect(res.text).toContain("1");
```
Comment (mmmaaatttttt) [inline __tests__/unit/integration/api.test.js:28]: can you add expectations on the status code for all of these as well?

File: __tests__/unit/integration/api.test.js [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+  });
+
+  test("batch request with range and specific number", async function () {
+    const res = await request(app).get("/1..3,10");
+    expect(res.body).toHaveProperty("1", "2", "3", "10");
+  });
+
+  test("invalid url for batch request", async function () {
+    const res = await request(app).get("/1...3");
+    expect(res.text).toEqual("Invalid url");
```
Comment (mmmaaatttttt) [inline __tests__/unit/integration/api.test.js:45]: re: my earlier comment, here's a case where i'd be interested in the status code. is this 400 level or 200 level? i feel like it should be the former, but maybe it's not. either way, making an assertion on the status code will help clarify.

File: __tests__/unit/integration/api.test.js [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+  test("batch request with type 'date'", async function () {
+    const res = await request(app).get("/3..5/date");
+    expect(res.body[3]).toContain("January 3rd");
+  });
+
+  test("batch request with type 'math'", async function () {
+    const res = await request(app).get("/3..5/math");
+    let resWords = res.body[3].split(" ");
+    expect(Number(resWords[0])).toEqual(3);
```
Comment (mmmaaatttttt) [inline __tests__/unit/integration/api.test.js:56]:
```suggestion
    expect(+resWords[0]).toEqual(3);
```

File: __tests__/unit/integration/api.test.js [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+  });
+
+  test("batch request with type 'trivia'", async function () {
+    const res = await request(app).get("/3..5/trivia");
+    let resWords = res.body[3].split(" ");
+    expect(Number(resWords[0])).toEqual(3);
```
Comment (mmmaaatttttt) [inline __tests__/unit/integration/api.test.js:62]:
```suggestion
    expect(+resWords[0]).toEqual(3);
```

File: __tests__/unit/integration/api.test.js [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+  });
+
+  test("batch request with type 'year'", async function () {
+    const res = await request(app).get("/3..5/year");
+    // has "is the year"
```
Comment (mmmaaatttttt) [inline __tests__/unit/integration/api.test.js:67]: you can remove this comment, the expectation is pretty clear

File: __tests__/unit/integration/api.test.js [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+  });
+
+  // tests for route "/:date([-0-9/.,]+)/:type(date)?"
+
+  // tests for route "/random/:type?"
+
+  test("random with without type defined", async function () {
+    const res = await request(app).get("/random");
+    let resWords = res.text.split(" ");
+    expect(typeof Number(resWords[0])).toEqual("number");
```
Comment (mmmaaatttttt) [inline __tests__/unit/integration/api.test.js:100]: i don't think this is testing anything, since `typeof Number(x)` will be "number" regardless of the value of x, no? Even if `Number(x)` is `NaN`, `typeof NaN` is (weirdly) "number". Can you find another way to express what you're trying to test here? It might be necessary to mock Math.random or something.

File: __tests__/unit/integration/api.test.js [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+  });
+
+  test("random with type 'date'", async function () {
+    const res = await request(app).get("/random/date");
+    let resWords = res.text.split(" ");
+    let months = [
```
Comment (mmmaaatttttt) [inline __tests__/unit/integration/api.test.js:106]: same here, maybe if you mocked Math.random you would be able to say with certainty which month comes back.

File: __tests__/unit/integration/api.test.js [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+      "August",
+      "September",
+      "October",
+      "November",
+      "December",
+    ];
+    expect(months).toContain(resWords[0]);
+  });
+
+  test("random with type 'math'", async function () {
```
Comment (mmmaaatttttt) [inline __tests__/unit/integration/api.test.js:123]: yeah same for all of these, i think mocking Math.random (or whatever is used to do the randomness) is the way to go here, since then these api calls will be deterministic.

File: routes/api.js [hunk 15]
```diff
@@ -1,6 +1,7 @@
 var _ = require("underscore");
 var fs = require("fs");
 var utils = require("../public/js/shared_utils.js");
+const { Console } = require("console");
```
Comment (mmmaaatttttt) [inline routes/api.js:4]: unused import, can you remove?

File: public/sass/screen.scss [hunk 16]
```diff
@@ -4,7 +4,6 @@
```
Comment (mmmaaatttttt) [inline public/sass/screen.scss:4]: this diff seems super noisy, did you actually change anything in this file? if not i'd revert.

**Review by mmmaaatttttt — COMMENTED** [review]
@WinnieGotYu @kathyn262 -- thanks for your work on this! mostly minor edits. biggest thing is for the routes involving randomness, I think we should be mocking whatever is doing the random selection so that the test output is deterministic and you can make explicit assertions on what you get back. (@stephenisau has done a lot of work on mocking for this project, so he might be able to help if you have questions on this)

**Comment by kathyn262** [comment]
@mmmaaatttttt thanks for reviewing! made the suggested changes!

**Review by mmmaaatttttt — APPROVED** [review]
